From 7a6e9489e3d8e3679f8f0fe7d2b445682f561f56 Mon Sep 17 00:00:00 2001
From: UN1CA <https://github.com/salvogiangri/UN1CA>
Date: Tue, 1 Sep 2026 15:32:41 +0200
Subject: [PATCH] Allow preloaded system app downgrade

Extend the existing UN1CA application downgrade toggle to the separate
preloaded-system-app version check. Without this, system app updates can
still fail with INSTALL_FAILED_VERSION_DOWNGRADE when the target APK is
older than the hidden package version from the system image.
---
 .../server/pm/PackageManagerService.smali        | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/smali_classes2/com/android/server/pm/PackageManagerService.smali b/smali_classes2/com/android/server/pm/PackageManagerService.smali
index f8bbdc69..5d8e2e2a 100644
--- a/smali_classes2/com/android/server/pm/PackageManagerService.smali
+++ b/smali_classes2/com/android/server/pm/PackageManagerService.smali
@@ -22845,6 +22845,22 @@
 
     if-nez v0, :cond_d
 
+    iget-object v0, v2, Lcom/android/server/pm/InstallPackageHelper;->mContext:Landroid/content/Context;
+
+    invoke-virtual {v0}, Landroid/content/Context;->getContentResolver()Landroid/content/ContentResolver;
+
+    move-result-object v0
+
+    const-string v1, "unica_allow_downgrade"
+
+    const/4 v3, 0x0
+
+    invoke-static {v0, v1, v3}, Landroid/provider/Settings$System;->getInt(Landroid/content/ContentResolver;Ljava/lang/String;I)I
+
+    move-result v0
+
+    if-nez v0, :cond_d
+
     :try_start_5
     invoke-interface {v14}, Lcom/android/server/pm/pkg/AndroidPackage;->getLongVersionCode()J
 
-- 
2.51.2
